src/coreclr/vm/marshalnative.cpp

```diff
@@ -176,68 +172,57 @@ extern "C" INT32 QCALLTYPE MarshalNative_SizeOfHelper(QCall::TypeHandle t, BOOL
     return rv;
 }
 
-
-/************************************************************************
- * PInvoke.OffsetOfHelper(Class, Field)
- */
-FCIMPL1(UINT32, MarshalNative::OffsetOfHelper, ReflectFieldObject *pFieldUNSAFE)
+extern "C" SIZE_T QCALLTYPE MarshalNative_OffsetOf(FieldDesc* pFD)
 {
     CONTRACTL
     {
-        FCALL_CHECK;
-        PRECONDITION(CheckPointer(pFieldUNSAFE));
+        QCALL_CHECK;
+        PRECONDITION(pFD != NULL);
     }
     CONTRACTL_END;
 
-    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    SIZE_T offset = 0;
 
-    FieldDesc *pField = refField->GetField();
-    TypeHandle th = TypeHandle(pField->GetApproxEnclosingMethodTable());
+    BEGIN_QCALL;
+
+    TypeHandle th = TypeHandle(pFD->GetApproxEnclosingMethodTable());
 
     if (th.IsBlittable())
     {
-        return pField->GetOffset();
+        offset = pFD->GetOffset();
     }
-
-    UINT32 externalOffset = 0;
-
-    HELPER_METHOD_FRAME_BEGIN_RET_1(refField);
+    else
     {
-        GCX_PREEMP();
-        // Determine if the type is marshalable.
+        // Verify the type can be marshalled.
         if (!IsStructMarshalable(th))
         {
-            // It isn't marshalable so throw an ArgumentException.
             StackSString strTypeName;
```

Comment:
```suggestion
            SString strTypeName;
```
We do not need stack-allocated buffer for exception throwing.